### ISSUE TITLE
fonction receive data with ucan 

### ISSUE THREAD

**Issue by SERIDJ** [idaniel86/systeccan#1]
hi please i need to reveive data from PCANview (USBCAN) with this library using **read_can_msg** and **read_cyclic_can_msg** but i dont know how i can use it ????????????

**Comment by idaniel86** [comment]
Hi, please have a look on [python-can](https://python-can.readthedocs.io/en/3.1.1/) library. This interface is already supported there. You can find docu and examples for working with CAN.

**Comment by SERIDJ** [comment]
i cant send data with python-can bcs i cant change Baudrate ????

**Comment by idaniel86** [comment]
import can

bus = can.interface.Bus(bustype='systec', channel=0, bitrate=?)

**Comment by SERIDJ** [comment]
yes yes i I have sucessfully done it but here i dont have event for receive i need to do while and using thread .....
but with library ucan i have **def can_msg_received_event(self, channel):** and when i print data i have **([<ucan.CanMsg object at 0x0335F9E0>], 0)** how i can decode it ?

**Comment by SERIDJ** [comment]
and my code is : 

    def can_msg_received_event(self, channel):
        """
        Event occurs when at leas one CAN message has been received.

        Call the method :meth:`read_can_msg` to receive the CAN messages.

        :param int channel: Specifies the CAN channel which received CAN messages.

        .. note:: To be overridden by subclassing.
        """
        lire_msg= self.read_can_msg(channel=Channel.CHANNEL_CH0,count=8)
           
      
        if lire_msg is not None:
            print(lire_msg)

**Comment by idaniel86** [comment]
The can message is a structure with properties, have a look in the source code for [CanMsg](https://github.com/idaniel86/systeccan/blob/b1a41dda37de236dd23784b3a5021461fe2a49f7/ucan.py#L632).

**Comment by SERIDJ** [comment]
i did bu i dont understand , u can tell me what is  structure please ?
 
class CanMsg(Structure):
    """
    Structure of a CAN message.

    .. seealso::

       :meth:`USBCanServer.read_can_msg`

       :meth:`USBCanServer.write_can_msg`

       :meth:`USBCanServer.define_cyclic_can_msg`

       :meth:`USBCanServer.read_cyclic_can_msg`
    """
    _pack_ = 1
    _fields_ = [
        ("m_dwID", DWORD),  # CAN Identifier
        ("m_bFF", BYTE),  # CAN Frame Format (see enum :class:`MsgFrameFormat`)
        ("m_bDLC", BYTE),  # CAN Data Length Code
        ("m_bData", BYTE * 8),  # CAN Data (array of 8 bytes)
        ("m_dwTime", DWORD,)  # Receive time stamp in ms (for transmit messages no meaning)
    ]

    def __init__(self, id=0, frame_format=MsgFrameFormat.MSG_FF_STD, data=[]):
        super(CanMsg, self).__init__(id, frame_format, len(data), (BYTE * 8)(*data), 0)

    @property
    def id(self): return self.m_dwID

    @id.setter
    def id(self, id): self.m_dwID = id

    @property
    def frame_format(self): return self.m_bFF

    @frame_format.setter
    def frame_format(self, frame_format): self.m_bFF = frame_format

    @property
    def data(self): return self.m_bData[:self.m_bDLC]

    @data.setter
    def data(self, data):
        self.m_bDLC = len(data)
        self.m_bData((BYTE * 8)(*data))

    @property
    def time(self): return self.m_dwTime

**Comment by SERIDJ** [comment]
yes i see it but i cant find structure 

red_msg = self..read_can_msg(channel=Channel.CHANNEL_CH0,count=8)

i know format is a tuple(list(CanMsg), int)

i try :  red_msg.[0] to access to to list and after i cant find data 

but i dont know how i can decode  it 
please u can help ?

**Comment by idaniel86** [comment]
In the previous message I have sent you direct link to the github source code for the CanMsg structure, what is there that you do not understand?

**Comment by SERIDJ** [comment]
what  i add to my code to print data and id in my   **def can_msg_received_event(self, channel):**

**Comment by idaniel86** [comment]
- id --> lire_msg.id

- data --> lire_msg.data

**Comment by SERIDJ** [comment]
no its not so easy u need to access to to list from tuple with lire_msg[0]  after u try to get get id and data but i have : AttributeError: 'list' object has no attribute 'id' and 
AttributeError: 'list' object has no attribute 'data'

**Comment by idaniel86** [comment]
Ok, the return value of `read_can_msg(self, channel, count)` is `tuple(list(CanMsg), int)`. So in your case call `lire_msg, _ = read_can_msg(channel=Channel.CHANNEL_CH0,count=8)`, where `lire_msg` is list of received messages that you acceess using indexes.

**Comment by SERIDJ** [comment]
yes but i have probleme  when i use print( lire_msg[0]) i have<ucan.CanMsg object at 0x03D7E170>

**Comment by idaniel86** [comment]
Then acces its properies, like `lire_msg[0].id` or `lire_msg[0].data`

**Comment by SERIDJ** [comment]
ah yes it work i try that without  **,_**  that why 
so the last question if u can help i use this library in my GUI interface so  copy def fonction event that not work and it work for disconnect_event and  connect_event !!!!


    def connect_event(self):
            self.text_can.setText('USB-CANmodul has been connected to the host.')
            self.text_can.adjustSize()
            
    def disconnect_event(self):
            self.text_can.setText('USB-CANmodul has been disconnected to the host.')
            self.text_can.adjustSize()
            
    def can_msg_received_event(self):
        lire_msg,_ = can.read_can_msg(channel=Channel.CHANNEL_CH0,count=8)
        if lire_msg is not None:
              print(lire_msg[0].data)

**Comment by idaniel86** [comment]
Sorry, I do not understand your question. What works and what does not?

**Comment by SERIDJ** [comment]
when i copy this event in my programme Gui and i use import ucan 
connect_event and disconnect_event work  when i connect  or disconnet module the fonction is executed 
but for can_msg_received_event does not work when i reveive data nothing is happen it work just in ucan library

**Comment by idaniel86** [comment]
Becasue you are not overriding the receive event function, its signature is `can_msg_received_event(self, channel)` and not `can_msg_received_event(self)`.

**Comment by SERIDJ** [comment]
no and ii dont have Error so if the event work i have error TypeError: can_msg_received_event() takes 1 positional argument but 2 were given

**Comment by idaniel86** [comment]
The code should be:
```
def can_msg_received_event(self, channel):
    lire_msg,_ = can.read_can_msg(channel=Channel.CHANNEL_CH0,count=8)
    if lire_msg is not None:
          print(lire_msg[0].data)
```

**Comment by SERIDJ** [comment]
yes yes but nothing is happen when i receive data the event not work for this fonction.
 it work only in library ucan

**Comment by SERIDJ** [comment]
and  connect_event et disconnect_event work in both : my programme GUI and library 
that why i dont understand

**Comment by idaniel86** [comment]
Sorry, I cannot help you here. You have to thinks this out for yourself. If your GUI class is derived from the USBCanServer, then overloading of the events should work.

**Comment by SERIDJ** [comment]
no its not derived from USBCanServer but i dont know why it work for donnect et disconnect ...

thnaks for ur help !

**Comment by idaniel86** [comment]
This is also mystery for me, if you want to use these events, you must derive your class from USBCanServer and overload these event callbacks.

Good luck

**Comment by SERIDJ** [comment]
ah sorry i use Inheritance  for USBCanServer  i forget 


class Window(QWidget,USBCanServer): 
    def __init__(self,left,top):
        super(Window, self).__init__()
        USBCanServer.__init__(self)

**Comment by idaniel86** [comment]
Ok, so then please try again this solution: https://github.com/idaniel86/systeccan/issues/1#issuecomment-494766917, where there is the `channel` input parameter.

**Comment by SERIDJ** [comment]
yes i did it not work and with python-can i dont know how to send Remote Request

**Comment by idaniel86** [comment]
So just please try a simple print:
```
def can_msg_received_event(self, channel):
    print("can_msg_received_event!")
```

**Comment by SERIDJ** [comment]
yes i did i have nothing in shell like the event not exist

**Comment by idaniel86** [comment]
And did you call `init_hardware`? This function is responsible for registering the event callbacks. The control callbacks like `connect_event` and `disconnect_event` are initialized in the constructor and therefore they allways work.

**Comment by SERIDJ** [comment]
yes i  do 

#***************************send data********************************    
can = USBCanServer()
try:
        can.init_hardware() #USB-CANmodul has been initialized
        can.init_can(channel=Channel.CHANNEL_CH0, BTR=Baudrate.BAUD_250kBit)#interface of an USB-CANmodul has been initialized
        print('USB-CANmodul has been initialized')

except :
       print('Cannot find USB-CANmodul .')
       #exit()

**Comment by idaniel86** [comment]
The problem is that you are creating 2 instances of `USBCanServer`, one that is `can` and the other is the instanse of `Window`. You are calling `init_hardware` for instance of `can` that has an empty `can_msg_received_event` callback. You have to do the whole inialization either inside the `Window` class or with its instance.
